[10.0] partner multi relation tabs migration

NL66278 opened the pull request.

NL66278 commented:
This PR depend on/includes https://github.com/OCA/partner-contact/pull/497

(Update 2018-04-13: PR 497 has now been merged)

hbrunn commented on partner_multi_relation/models/res_partner_relation_all.py:
```diff
@@ -1,18 +1,100 @@
+    date_end: end date of relation if relevant or NULL,
+    %(is_inverse)s as is_inverse: used to determine type_selection_id,
+    """
+    global _last_key_offset
```
this will fail for multithreaded servers. We need to think of another mechanism, I'd suggest splitting up the sql statement into function similar to https://github.com/OCA/OCB/blob/10.0/addons/account/report/account_invoice_report.py, then overrides can add what they need without destroying multithreading

NL66278 replied:
I don't think it can fail as the variable is only used during module loading to gather all the parts of the statement to create the view. I would be amazed if the loading of one module could be interrupted by the loading of anpther module.

hbrunn reviewed (changes requested):
test coverage needs to be increased

NL66278 commented:
@hbrunn Tests have been added, so I think your point is satisfied now.

hbrunn commented:
thanks!

NL66278 committed:
[IMP] New module partner_multi_relation_tabs.

NL66278 committed:
[IMP] partner_multi_relations_tabs. Clean code principles.

- Separate model (db) partially from user interface;
- Use demo data for performance, cleaner tests, and front end testing;
- Make sure tabs dependent on partner_category_id work.

NL66278 committed:
[IMP] partner_multi_relation_tabs. Forward to 100% coverage.

NL66278 commented:
@ddufresne Would you be so kind as to have a look at this PR? I also intent to migrate it to 11.0.

ddufresne commented:
@NL66278 this is quite a lot of code. Before I go deep into review, I would prefer to see the big picture.

Could you extend your README with screenshots and explanations on how to use your module step by step. Then, it's going to be easier for me to review.

NL66278 committed:
[IMP] partner_multi_relation_tabs. Add USAGE.rst with screenshots.

NL66278 commented:
@ddufresne I added a usage section with screenshots (taking as an example your code for partner_multi_relation in 12.0). However the images will only be rendered properly from the README.rst after the merge, because of the way the url's have been setup. But of course they are part of the code so you should be able to see them there.

NL66278 committed:
[FIX] Do not add tabs on browse - to costly.

NL66278 committed:
[FIX] partner_multi_relation_tab. Prevent key errors.

There are situations where tabs are added by fields_view_get, but not yet visible for other functions,
presumably in other workers. Handle the errors caused by this. So far only seen in mapped function.

lfreeke reviewed (approved):
Functional test :+1:

hbrunn commented:
/ocabot merge
